test/reputation-mining/disputes-over-child-reputation.js

```diff
@@ -218,14 +222,346 @@ contract("Reputation Mining - disputes over child reputation", accounts => {
+
+      await setupFinalizedTask({
+        colonyNetwork,
+        colony: metaColony,
+        skillId: 5,
+        managerPayout: 100000000000,
+        evaluatorPayout: 100000000,
+        workerPayout: 500000000000,
+        managerRating: 1,
+        workerRating: 1,
+        worker: MINER2
+      });
+
+      await goodClient.resetDB();
+      await advanceMiningCycleNoContest({ colonyNetwork, test: this, client: goodClient });
+      await goodClient.saveCurrentState();
+
+      // The update log should contain the person being rewarded for the previous
+      // update cycle, and reputation updates for one task completion (manager, worker, evaluator);
+      // That's five in total.
+      const repCycle = await getActiveRepCycle(colonyNetwork);
+      const nLogEntries = await repCycle.getReputationUpdateLogLength();
+      assert.equal(nLogEntries.toNumber(), 5);
+
+      const badClient = new MaliciousReputationMinerExtraRep(
+        { loader, realProviderPort, useJsTree, minerAddress: MINER2 },
+        24, // Passing in update number for colony wide skillId: 5, user: 0
+        "0xfffffffffffffffffffffff"
+      );
+
+      // Moving the state to the bad client
+      await badClient.initialise(colonyNetwork.address);
+      const currentGoodClientState = await goodClient.getRootHash();
+      await badClient.loadState(currentGoodClientState);
+
+      await submitAndForwardTimeToDispute([goodClient, badClient], this);
+
+      // Run through the dispute until we can call respondToChallenge
+      await goodClient.confirmJustificationRootHash();
+      await badClient.confirmJustificationRootHash();
+      await runBinarySearch(goodClient, badClient);
+      await goodClient.confirmBinarySearchResult();
+      await badClient.confirmBinarySearchResult();
 
+      // Now get all the information needed to fire off a respondToChallenge call
+      const [round, index] = await goodClient.getMySubmissionRoundAndIndex();
+      const submission = await repCycle.getDisputeRounds(round.toString(), index.toString());
+      const firstDisagreeIdx = new BN(submission[8].toString());
+      const lastAgreeIdx = firstDisagreeIdx.subn(1);
+      const reputationKey = await goodClient.getKeyForUpdateNumber(lastAgreeIdx.toString());
+      const [agreeStateBranchMask, agreeStateSiblings] = await goodClient.justificationTree.getProof(`0x${lastAgreeIdx.toString(16, 64)}`);
+      const [disagreeStateBranchMask, disagreeStateSiblings] = await goodClient.justificationTree.getProof(`0x${firstDisagreeIdx.toString(16, 64)}`);
+      const logEntryNumber = await goodClient.getLogEntryNumberForLogUpdateNumber(lastAgreeIdx.toString());
+      const lastAgreeJustifications = goodClient.justificationHashes[`0x${lastAgreeIdx.toString(16, 64)}`];
+      const firstDisagreeJustifications = goodClient.justificationHashes[`0x${firstDisagreeIdx.toString(16, 64)}`];
+
+      await checkErrorRevert(
+        repCycle.respondToChallenge(
```

Comment:
Yeah, this can go either way. If you'd rather see these in custom mining clients, happy to do that. I have no strong feelings either way.

Reply:
Seems like there are 6 instances of this approach in tests (all in `dispute-resolution-misbehaviour` file). It appears justifiable to have a generic bad client that fakes `respondToChallenge` for a given parameter.

Reply:
Also appears that the following bad clients are no longer used, if you can kindly remove them please
`MaliciousReputationMiningClaimNoUserChildReputation`
`MaliciousReputationMiningWrongChildReputation`
`MaliciousReputationMiningWrongOriginReputation`
`MaliciousReputationMiningWrongNewestReputation`
`MaliciousReputationMiningWrongProofLogEntry` (which already does a bad `respondToChallenge` btw)

Reply:
These clients are used, but there was a Mining/Miner inconsistency between the files themselves and where they were used / how they were imported. I *think* I've straightened all that out now (should now be 'Miner' everywhere).

I have added a new bad client that replaces an indicated parameter in `respondToChallenge` with a new value. I have not replaced `MaliciousReputationMinerWrongProofLogEntry` with this in the test where it is used because that client adjusts the valid value by a given amount, rather than just replacing it with a known wrong value. This replacement is certainly possible, but the test in question becomes more brittle if we do so.